File: Binary Tree/108_Convert_Sorted_Array_to_Binary_SearchTree.py

# Definition for a binary tree node.

class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

def sortedArrayToBST(self, arr):
    """
    :type arr: List[int]
    :rtype: TreeNode
    """
    mid = (len(arr))//2
    root = TreeNode(arr[mid])
    root.left = sortedArrayToBST(arr[:mid])
    root.right = sortedArrayToBST(arr[mid+1:])
    return root

File: Binary Tree/108_Convert_Sorted_Array_to_Binary_SearchTree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {number[]} nums
 * @return {TreeNode}
 * 題目，給一組已經排序好的array, 回傳一組balanced的binary tree,
 * 解法，用recursion來實作，每一個節點的值都是傳進來的陣列中間那個元素，他的左邊子節點則是陣列左邊一半的中心點
 * 右邊子節點也是右邊陣列的中心點，如此類推，一直到傳進去的陣列的長度爲0爲止。
 * 用slice有點浪費，其實可以用index來追縱即可。
 */

function TreeNode(val) {
    this.val = val;
    this.left = this.right = null;
}

var sortedArrayToBST = function(nums) {

    if (!nums.length) {
        return null;
    }
    let mid = Math.floor(nums.length / 2);
    let root = new TreeNode(nums[mid]);

    root.left = sortedArrayToBST(nums.slice(0, mid));
    root.right = sortedArrayToBST(nums.slice(mid + 1, nums.length));

    return root;
};

console.log(
    "sortedArrayToBST([-10,-3,0,5,9])",
    sortedArrayToBST([-10, -3, 0, 5, 9]),
);


